src/app/libp2p_helper/src/libp2p_helper/config_msg.go

```diff
@@ -144,6 +160,119 @@ func configurePubsub(app *app, validationQueueSize int, directPeers []peer.AddrI
 	return err
 }
 
+func readTopicLevels(lst ipc.TopicLevel_List) (topicLevelConf, error) {
+	res := make(topicLevelConf)
+	for i := 0; i < lst.Len(); i++ {
+		tl := lst.At(i)
+		topics, err := tl.Topics()
+		if err != nil {
+			return nil, err
+		}
+		for j := 0; j < topics.Len(); j++ {
+			topic, err := topics.At(j)
+			if err != nil {
+				return nil, err
+			}
+			l, has := res[topic]
+			if has {
+				if i == l.maxLevel+1 {
+					res[topic] = topicLevelEntry{
+						minLevel: l.minLevel,
+						maxLevel: i,
+					}
+				} else {
+					return nil, fmt.Errorf("duplicate topic on level %d: %s (minL: %d, maxL: %d)", i, topic, l.minLevel, l.maxLevel)
+				}
+			} else {
+				res[topic] = topicLevelEntry{minLevel: i, maxLevel: i}
+			}
+		}
+	}
+	return res, nil
+}
+
+type topicLevelConf map[string]topicLevelEntry
+
+type topicLevelEntry struct {
+	minLevel int
+	maxLevel int
+}
+
+// leveledSubscriptionFilter is a filter that takes topic level parameter
+// which lists topics for each level.
+// Then, for each peer we monitor topics being used and discard
+// subscriptions from topics of lower level than the max level
+// observed for the node.
+// Filter keeps a record per each connected node that subscribed to some topic,
+// and cleans up the record when the node disconnects.
+type leveledSubscriptionFilter struct {
+	topicLevel topicLevelConf
+	peerLevel  map[peer.ID]int
+	lock       sync.RWMutex
+	localPeer  peer.ID
+}
+
+func (f *leveledSubscriptionFilter) CanSubscribe(topic string) bool {
+	_, has := f.topicLevel[topic]
+	return has
+}
+func (f *leveledSubscriptionFilter) FilterIncomingSubscriptions(pid peer.ID, subs []*pb.RPC_SubOpts) ([]*pb.RPC_SubOpts, error) {
+	f.lock.RLock()
+	pl, hasPl := f.peerLevel[pid]
+	f.lock.RUnlock()
+	initHasPl := hasPl
+	peerLevel := pl
+	for _, sub := range subs {
+		l, hasL := f.topicLevel[sub.GetTopicid()]
+		if !hasL {
+			continue
+		}
+		if !hasPl || pl < l.minLevel {
+			hasPl = true
+			pl = l.minLevel
+		}
+	}
+	if !hasPl {
+		// all subs are irrelevant
+		return nil, nil
+	}
+	if !initHasPl || peerLevel < pl {
+		f.lock.Lock()
+		peerLevel, initHasPl = f.peerLevel[pid]
+		if !initHasPl || peerLevel < pl {
+			f.peerLevel[pid] = pl
+			peerLevel = pl
+		}
+		f.lock.Unlock()
+	}
+	res := pubsub.FilterSubscriptions(subs, func(topic string) bool {
+		l, hasL := f.topicLevel[topic]
+		return hasL && l.minLevel <= peerLevel && peerLevel <= l.maxLevel
```

Comment:
This level abstraction feels very prescriptive about which bundles of topics other nodes have to support. Once things are split up, won't we be in a world where some implementers might choose to upgrade e.g. the snark work subscription first, while others might choose to upgrade e.g. the transaction subscription first?